### PR TITLE
Fetch ML data

### PR DESCRIPTION
**Pull request by lperryp**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/25247

## Description
An automation for fetching ML data

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 4.5.0
- [ ] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 1]
```diff
@@ -13,7 +1038,17 @@ def return_json_entry(obj):
 
 
 def main():
-    return_json_entry({})
+    incidents_query_args = demisto.args()
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:1041]: default modified > last run time
first run - run on last 6 months with max of 1,000-3,000
we can store the state in list: state will store last run time

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 2]
```diff
@@ -13,7 +1038,17 @@ def return_json_entry(obj):
...
+    incidents_query_res = demisto.executeCommand('GetIncidentsByQuery', incidents_query_args)
+    if is_error(incidents_query_res):
+        return_error(get_error(incidents_query_res))
+    incidents = json.loads(incidents_query_res[-1]['Contents'])
+    data = extract_data_from_incidents(incidents)
+    encoded_data = json.dumps(data).encode('utf-8', errors='ignore')
+    compressed_data = zlib.compress(encoded_data, 4)
+    compressed_hr_data = b64encode(compressed_data).decode('utf-8')
+    res = {'PayloadVersion': FETCH_DATA_VERSION, 'PayloadData': compressed_hr_data}
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:1050]: check what is the payload size (length)

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 3]
```diff
@@ -1,6 +1,1031 @@
...
+     "adnld.com", "adshor.tk", "adspl.us", "adzip.us", "articleshrine.com", "asso.in", "b2s.me", "bih.cc", "biturl.net",
+     "buraga.org", "cc.cr", "cf6.co", "dollarfalls.info", "domainonair.com", "gooplu.com", "hide4.me", "ik.my",
+     "ilikear.ch", "infovak.com", "itz.bz", "jetzt-hier-klicken.de", "kly.so", "lst.bz", "mrte.ch",
+     "multiurlscript.com", "nowlinks.net", "nsyed.com", "ooze.us", "ozn.st", "scriptzon.com", "short2.in",
+     "shortxlink.com", "shr.tn", "shrt.in", "sitereview.me", "sk.gy", "snpurl.biz", "socialcampaign.com", "swyze.com",
+     "theminiurl.com", "tinylord.com", "tinyurl.ms", "tip.pe", "ty.by"])
+
+
+def find_label_fields_candidates(incidents_df):
+    candidates = [col for col in list(incidents_df) if
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:651]: look on the last 500 incidents (sorted by "created" time)
filter empty fields (empty in more than 50%)

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 4]
```diff
@@ -1,6 +1,1031 @@
...
+                  sum(isinstance(x, str) for x in incidents_df[col]) > 0.5 * len(incidents_df)]
+    candidates = [col for col in candidates if col not in LABEL_FIELDS_BLACKLIST]
+
+    candidate_to_values = {col: incidents_df[col].unique() for col in candidates}
+    candidate_to_values = {col: values for col, values in candidate_to_values.items() if
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:656]: consider empty values, we can give up this I think

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 5]
```diff
@@ -1,6 +1,1031 @@
...
+                              'category', 'closeNotes', 'closed', 'closingUserId', 'created', 'criticalassets',
+                              'dbotCreatedBy', 'details', 'detectionsla', 'droppedCount', 'dueDate', 'emailbody',
+                              'emailheaders', 'emailsubject', 'hasRole', 'id', 'investigationId', 'isPlayground',
+                              'labels', 'lastJobRunTime', 'lastOpen', 'linkedCount', 'linkedIncidents', 'modified',
+                              'name', 'notifyTime', 'occurred', 'openDuration', 'owner', 'parent', 'phase',
+                              'playbookId', 'previousRoles', 'rawCategory', 'rawCloseReason', 'rawJSON', 'rawName',
+                              'rawPhase', 'rawType', 'reason', 'remediationsla', 'reminder', 'roles', 'runStatus',
+                              'severity', 'sla', 'sortValues', 'sourceBrand', 'sourceInstance',
+                              'status', 'timetoassignment', 'type', 'urlsslverification', 'version'])
+LABEL_VALUES_KEYWORDS = ['spam', 'malicious', 'legit', 'false', 'positive', 'phishing', 'fraud', 'internal', 'test']
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:615]: add tp, fp, resolve, credentials, spare, malware
maybe more phishing types?

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 6]
```diff
@@ -1,6 +1,1031 @@
...
+    candidate_to_values = {col: values for col, values in candidate_to_values.items() if
+                           all(isinstance(v, str) for v in values)}
+
+    # filter columns by unique values count
+    candidate_to_values = {col: values for col, values in candidate_to_values.items() if 0 < len(values) < 15}
+    candidate_to_values = {col: [v.lower() for v in values] for col, values in candidate_to_values.items()}
+    candidate_to_score = {col: 0 for col in candidate_to_values}
+    for col, values in candidate_to_values.items():
+        for v in values:
+            if any(w in v for w in LABEL_VALUES_KEYWORDS):
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:665]: consider to check also the field name, boolean
if the field contains:  "classifi", "type", "resolution",  "reason", "category", "disposition", "severity", "malicious", "tag" add more fields

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 7]
```diff
@@ -1,6 +1,1031 @@
...
+    email_attachments = email_attachments if email_attachments is not None else []
+    if email_body is None or email_body.strip() == '' or isinstance(email_body, float):
+        email_body = get_text_from_html(email_html)
+    if isinstance(email_html, float):
+        email_html = ''
+    if isinstance(email_subject, float):
+        email_subject = ''
+    email_body, email_subject = email_body.strip().lower(), email_subject.strip().lower()
+    text = email_subject + ' ' + email_body
+    email_body_word_tokenized = word_tokenize(email_body)
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:951]: filter if email text is empty or length < 20 chars

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 8]
```diff
@@ -1,6 +1,1031 @@
...
+            all_attachments_names.append(attachment_name.lower())
+    all_attachments_names_lengths = [len(name) for name in all_attachments_names]
+    res['min_attachment_name_length'] = min(all_attachments_names_lengths) if len(
+        all_attachments_names_lengths) > 0 else 0
+    res['max_attachment_name_length'] = max(all_attachments_names_lengths) if len(
+        all_attachments_names_lengths) > 0 else 0
+    res['avg_attachment_name_length'] = float(sum(all_attachments_names_lengths)) / len(  # type: ignore
+        all_attachments_names_lengths) if len(all_attachments_names_lengths) > 0 else 0  # type: ignore
+    res['image_extension'] = 0
+    for image_format in ['.jepg', '.gif', '.bmp', '.png']:
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:907]: put it in CONST var
missing jpg
fix jpeg
more image extension
bring the extension

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 9]
```diff
@@ -1,6 +1,1031 @@
...
+    for c in string.printable:
+        if c in ['[', ']', '<']:
+            continue
+        count = text.count(c)
+        if count > 0:
+            characters_count[c] = count
+    return characters_count
+
+
+def get_url_features(email_body, email_html, soup):
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:742]: check extract domain automation \ extract url

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 10]
```diff
@@ -1,6 +1,1031 @@
...
+def get_html_features(soup):
+    global HTML_TAGS
+    html_counter = Counter([tag.name for tag in soup.find_all()])
+    for t in HTML_TAGS:
+        html_counter[t] = 0 if t not in html_counter else html_counter[t]
+    html_counter = {k: v for k, v in html_counter.items() if k in HTML_TAGS}
+    return html_counter
+
+
+def get_ml_features(tokenized_text):
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:777]: change to embedding

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 11]
```diff
@@ -1,6 +1,1031 @@
...
+    global EMBEDDING_DICT
+    if EMBEDDING_DICT is None:
+        with open('/var/glove_top_20k.p', 'rb') as file:
+            EMBEDDING_DICT = pickle.load(file)
+    vectors = [EMBEDDING_DICT[w] for w in tokenized_text if w in EMBEDDING_DICT]
+    if len(vectors) == 0:
+        mean_vector = np.zeros(50)
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:784]: maybe bring also 100

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 12]
```diff
@@ -1,6 +1,1031 @@
...
+'''
+Define vocabulary to search
+'''
+
+word_to_regex = {
+    "x days": r"\d+ days",
+    "x hours": r"(\d+ hours)|(\d+ hrs)|(\d+hrs)",
+}
+
+word_to_ngram = {
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:55]: consider to work with "set" and existing word

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 13]
```diff
@@ -1,6 +1,1031 @@
...
+    # break multi-headlines into a line each
+    chunks = (phrase.strip() for line in lines for phrase in line.split("  "))
+    # drop blank lines
+    text = '\n'.join(chunk for chunk in chunks if chunk)
+    return text
+
+
+def get_vocab_features(text, ngrams_counter):
+    global word_to_ngram, word_to_regex
+    vocav_features = {}
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:693]: rename: vocab

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 14]
```diff
@@ -1,6 +1,1031 @@
...
+            vocav_features[word] = count
+    for word, ngram in word_to_ngram.items():
+        count = ngrams_counter[ngram] if ngram in ngrams_counter else 0
+        if count > 0:
+            vocav_features[word] = count
+    return vocav_features
+
+
+def get_lexical_features(email_subject, email_body, email_body_word_tokenized, email_subject_word_tokenized):
+    words = [w for w in email_body_word_tokenized if any(c.isalnum() for c in w)]
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:706]: maybe will also add number of non-ascii characters? money characters?

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 15]
```diff
@@ -1,6 +1,1031 @@
...
+        'all_urls_count': len(all_urls),
+        'average_url_length': average_url_length,
+        'min_url_length': min_url_length,
+        'max_url_length': max_url_length,
+        'shortened_urls_count': shortened_urls_count,
+        'drive_count': drive_count
+    }
+
+
+def get_html_features(soup):
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:773]: I will also the number of URLs inside the email body that the text different than the URL itself
(like "click here" where here is a link)

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 16]
```diff
@@ -1,6 +1,1031 @@
...
+    else:
+        return headers_with_name[index]['value']
+
+
+def get_headers_features(email_headers):
+    spf_result = get_header_value(email_headers, header_name='Received-SPF')
+    if spf_result is not None:
+        spf_result = spf_result.split()[0].lower()
+    else:
+        spf_result = 'other'
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:813]: I will say "empty" not other - if the header doesn't exists
also please use const not magic string

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 17]
```diff
@@ -1,6 +1,1031 @@
...
+        res_spf[spf_result] += 1
+    else:
+        res_spf['other'] += 1
+    res_spf['non-positive'] = spf_result in set(['none', 'fail', 'softfail'])
+    authentication_results = get_header_value(email_headers, header_name='Authentication-Results')
+    if authentication_results is not None:
+        if 'dkim=' in authentication_results:
+            dkim_result = authentication_results.split('dkim=')[-1].split()[0].lower().strip()
+        else:
+            dkim_result = 'other'
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:825]: 'empty' const

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 18]
```diff
@@ -1,6 +1,1031 @@
...
+    else:
+        dkim_result = 'other'
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:827]: 'empty' const

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 19]
```diff
@@ -1,6 +1,1031 @@
...
+        res_dkim[dkim_result] += 1
+    else:
+        res_dkim['other'] += 1
+    res_dkim['non-positive'] = spf_result in set(['none', 'fail', 'softfail'])
+    list_unsubscribe = get_header_value(email_headers, header_name='List-Unsubscribe')
+    unsubscribe_post = get_header_value(email_headers, header_name='List-Unsubscribe-Post-SPF')
+    res_unsubscribe = 1 if list_unsubscribe is not None or unsubscribe_post is not None else 0
+    from_value = get_header_value(email_headers, header_name='From')
+    return_path_value = get_header_value(email_headers, header_name='Return-Path')
+    last_received_value = get_header_value(email_headers, header_name='Received', index=-1)
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:839]: add also "reply-to" header and compare to the from address (I think envelope sender represent by the return path so you already extract that)
also maybe worth to extract the domain  for each one to compare between them

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 20]
```diff
@@ -1,6 +1,1031 @@
...
+    if last_received_value is not None:
+        received_address_list = re.findall(r'(?<=from )\S+', last_received_value)
+        if len(received_address_list) > 0:
+            received_address = received_address_list[0]
+        else:
+            received_address = ''
+    else:
+        received_address = ''
+    try:
+        from_domain = from_address.split('@')[-1].split('.')[-2] if from_address != '' else ''
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:857]: create  function that extract domain

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 21]
```diff
@@ -1,6 +1,1031 @@
...
+        res_dkim['other'] += 1
+    res_dkim['non-positive'] = spf_result in set(['none', 'fail', 'softfail'])
+    list_unsubscribe = get_header_value(email_headers, header_name='List-Unsubscribe')
+    unsubscribe_post = get_header_value(email_headers, header_name='List-Unsubscribe-Post-SPF')
+    res_unsubscribe = 1 if list_unsubscribe is not None or unsubscribe_post is not None else 0
+    from_value = get_header_value(email_headers, header_name='From')
+    return_path_value = get_header_value(email_headers, header_name='Return-Path')
+    last_received_value = get_header_value(email_headers, header_name='Received', index=-1)
+    if from_value is not None:
+        from_address = re.findall(r'<\S+>', from_value)[0][1:-1]
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:841]: please use regex pattern - compile it once

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 22]
```diff
@@ -1,6 +1,1031 @@
...
+    else:
+        from_address = ''
+    if return_path_value is not None:
+        return_path_adress = return_path_value.strip().lstrip('<').rstrip('>').strip()
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:845]: you can use the regex pattern here too to extract the email, maybe it's safer?

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 23]
```diff
@@ -1,6 +1,1031 @@
...
+        if len(received_address_list) > 0:
+            received_address = received_address_list[0]
+        else:
+            received_address = ''
+    else:
+        received_address = ''
+    try:
+        from_domain = from_address.split('@')[-1].split('.')[-2] if from_address != '' else ''
+    except Exception:
+        print('error on from_address: {}'.format(from_address))
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:859]: don't print in the automation, also don't collect it to any logs,cuz it contains the email address

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 24]
```diff
@@ -1,6 +1,1031 @@
...
+        from_domain = ''
+
+    if from_address == '':
+        return_path_same_as_from = float('nan')
+    else:
+        return_path_same_as_from = from_address == return_path_adress
+
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:866]: add some more features as mentioned
return path domain as from domain
reply to as from
reply to domain as from
reply to as return path
reply to domain as return path domain
reply to exists

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 25]
```diff
@@ -1,6 +1,1031 @@
...
+        from_domain_with_received = float('nan')
+    else:
+        from_domain_with_received = from_domain in received_address
+    received_res = {}
+    recevied_headers = [header_dict for header_dict in email_headers if header_dict['name'] == 'Received']
+    count_received = len(recevied_headers)
+    received_res['count_received'] = count_received
+    content_type_value = get_header_value(email_headers, header_name='Content-Type', index=0, ignore_case=True)
+    if content_type_value is not None and ';' in content_type_value:
+        content_type_value = content_type_value.split(';')[0]
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:877]: maybe get all the content-types and not just the first? if it's possible

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 26]
```diff
@@ -1,6 +1,1031 @@
...
+    vectors = [EMBEDDING_DICT[w] for w in tokenized_text if w in EMBEDDING_DICT]
+    if len(vectors) == 0:
+        mean_vector = np.zeros(50)
+    else:
+        mean_vector = np.mean(vectors, axis=0)
+    res = {'glove_avg_{}'.format(i): mean_vector[i].item() for i in range(len(mean_vector))}
+    return res
+
+
+def get_header_value(email_headers, header_name, index=0, ignore_case=False):
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:796]: I think it terms of efficiency - it's worth to build email headers dict (key->value) once in `get_headers_features` method and pass that every time to this function 
here you iterate the headers list everytime you get header

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 27]
```diff
@@ -1,6 +1,1031 @@
...
+            continue
+        count = text.count(c)
+        if count > 0:
+            characters_count[c] = count
+    return characters_count
+
+
+def get_url_features(email_body, email_html, soup):
+    http_regex = r'http://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+'
+    https_regex = r'https://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+'
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:749]: if you stay with your regex (might be easier) please add support with `hxxp` it's proofpoint format

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 28]
```diff
@@ -1,6 +1,1031 @@
...
+    https_urls = re.findall(https_regex, email_body)
+    embedded_urls = [a['href'] for a in soup.findAll('a')]
+    all_urls = http_urls + https_urls + embedded_urls
+    all_urls_lengths = [len(u) for u in all_urls]
+    average_url_length = 0 if len(all_urls) == 0 else sum(all_urls_lengths) / len(all_urls)
+    min_url_length = 0 if len(all_urls_lengths) == 0 else min(all_urls_lengths)
+    max_url_length = 0 if len(all_urls_lengths) == 0 else max(all_urls_lengths)
+    shortened_urls_count = len([u for u in all_urls if any(shortened_u in u for shortened_u in SHORTENED_DOMAINS)])
+    drive_count = len([u for u in all_urls if any(drive in u for drive in DRIVE_URL_KEYWORDS)])
+    return {
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:760]: add also maybe the number of "hidden" links - where the text different than the link itself

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 29]
```diff
@@ -1,6 +1,1031 @@
...
+        'http_urls_count': len(http_urls),
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:761]: more possible features:
- compare the URL domain to the sender domain & return path domain 
- alexa rank for each one of the domains (you can return it as list)
- list of TLDs of the domain

There is a python package "tldextract" that might be useful here with extracting the domain & tld

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 30]
```diff
@@ -1,6 +1,1031 @@
...
+    for k, v in res_spf.items():
+        res['spf::{}'.format(k)] = v
+    for k, v in res_dkim.items():
+        res['dkim::{}'.format(k)] = v
+    res['unsubscribe_headers'] = res_unsubscribe
+    for k, v in adress_res.items():
+        res[k] = v  # type: ignore
+    for k, v in received_res.items():
+        res[k] = v
+    res['content-type::{}'.format(content_type_value)] = 1
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:892]: more possible features:
- from domain alexa rank
- return path domain alexa rank
- reply to alexa rank

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 31]
```diff
@@ -1,6 +1,1031 @@
...
+        if attachment_name is not None:
+            all_attachments_names.append(attachment_name.lower())
+    all_attachments_names_lengths = [len(name) for name in all_attachments_names]
+    res['min_attachment_name_length'] = min(all_attachments_names_lengths) if len(
+        all_attachments_names_lengths) > 0 else 0
+    res['max_attachment_name_length'] = max(all_attachments_names_lengths) if len(
+        all_attachments_names_lengths) > 0 else 0
+    res['avg_attachment_name_length'] = float(sum(all_attachments_names_lengths)) / len(  # type: ignore
+        all_attachments_names_lengths) if len(all_attachments_names_lengths) > 0 else 0  # type: ignore
+    res['image_extension'] = 0
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:911]: please bring all the extensions as list

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py (hunk 15 above)
Comment (lperryp) [reply Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:792]: it's counted as "embedded_urls" under url_features

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py (hunk 26 above)
Comment (lperryp) [reply Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:854]: reason for that is mainly the for the Received headers which appear multiple times

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 32]
```diff
@@ -740,21 +785,23 @@ def get_characters_features(text):
...
-    embedded_urls = [a['href'] for a in soup.findAll('a')]
-    all_urls = http_urls + https_urls + embedded_urls
+    url_regex = r'(?:(?:https?|ftp|hxxps?):\/\/|www\[?\.\]?|ftp\[?\.\]?)(?:[-\w\d]+\[?\.\]?)+[-\w\d]+(?::\d+)?' \
+                r'(?:(?:\/|\?)[-\w\d+&@#\/%=~_$?!\-:,.\(\);]*[\w\d+&@#\/%=~_$\(\);])?'
+    embedded_urls = []
+    for a in soup.findAll('a'):
+        if a.has_attr('href'):
+            embedded_urls.append(a['href'])
+    plain_urls = re.findall(url_regex, email_body)
+    all_urls = plain_urls + embedded_urls
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:795]: I think sometimes you can count url twice if it's HTML and also the text is the URL itself, maybe you should dedup this list using set

File: Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py [hunk 33]
```diff
@@ -740,21 +785,23 @@ def get_characters_features(text):
...
-    http_urls = re.findall(http_regex, email_body)
-    https_urls = re.findall(https_regex, email_body)
-    embedded_urls = [a['href'] for a in soup.findAll('a')]
-    all_urls = http_urls + https_urls + embedded_urls
+    url_regex = r'(?:(?:https?|ftp|hxxps?):\/\/|www\[?\.\]?|ftp\[?\.\]?)(?:[-\w\d]+\[?\.\]?)+[-\w\d]+(?::\d+)?' \
+                r'(?:(?:\/|\?)[-\w\d+&@#\/%=~_$?!\-:,.\(\);]*[\w\d+&@#\/%=~_$\(\);])?'
+    embedded_urls = []
+    for a in soup.findAll('a'):
+        if a.has_attr('href'):
+            embedded_urls.append(a['href'])
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py:793]: maybe check if the HTML text different than the URL itself

**Comment by lperryp** [comment]
@kirbles19 - this automation is for internal use for telemetry and won't be used by customers, so that can't be checked briefly (if at all)

**Comment by lperryp** [comment]
made some performance checks [here](https://docs.google.com/spreadsheets/d/13DQ8FwCtxy_yJlDKbXWYGSY6o2AAdi3zebXmUstxKds/edit?usp=sharing) (TL;DR: preprocessing time of an incident ranges 0.005-0.1 sec, average is 0.005). 
Also couldn't find an error or anything irregular at the logs (attached) 
[logs-bundle-02Aug2011_34UTC.tar.gz](https://github.com/demisto/content/files/5012073/logs-bundle-02Aug2011_34UTC.tar.gz)

@erezh31

**Comment by Itay4** [comment]
@lperryp please note the lgtm alert

**Comment by lperryp** [comment]
> @lperryp please note the lgtm alert

thanks @Itay4

**Comment by lperryp** [comment]
@Itay4 can we merge this? it fails on secrets detection which I tried to disable on the script

**Pull request merged**